### PR TITLE
add update User Profile

### PR DESCRIPTION
**Pull request by Tanyaa-a**
## Description
-- add Update functionality to UserController and Route to update

## Screenshots

Screenshot 2024-11-19 at 11 32 04 PM

### CODE REVIEW

File: src/controllers/userController.js [hunk 1]
```diff
@@ -1,28 +1,31 @@
...
-
-        const { password, _id, __v, ...userData } = user.toObject();
-        
-        res.status(StatusCodes.CREATED).json({ userData, token });
-    } catch (error) {
-      console.error(error);
-      res.status(StatusCodes.BAD_REQUEST).json({ error: error.message });
-      }
+  try {
+    const { email } = req.body;
```
Comment (benochi) [inline src/controllers/userController.js:8]: we can check if there is email in the req.body and immediately return a validation check.

File: src/controllers/userController.js [hunk 2]
```diff
@@ -59,8 +64,43 @@ const logout = async (req, res) => {
...
+    }
+
+    const decoded = jwt.verify(token, process.env.JWT_SECRET);
+    const userId = decoded.userId;
+
+    const { firstName, lastName, location } = req.body;
+    if (!firstName && !lastName && !location) {
+      throw new BadRequestError('No valid fields to update');
+    }
+
```
Comment (benochi) [inline src/controllers/userController.js:99]: Make sure that undefined doesnt bypass this if check, or they could get set to undefined in the DB.: 
if (firstName !== undefined) updates.firstName = firstName;
if (lastName !== undefined) updates.lastName = lastName;
if (location !== undefined) updates.location = location;

if (Object.keys(updates).length === 0) {
  throw new BadRequestError('No valid fields to update');
}

const updatedUser = await User.findByIdAndUpdate(
  userId,
  { $set: updates },
  { new: true, runValidators: true }
);

File: src/controllers/userController.js [hunk 3]
```diff
@@ -32,21 +35,23 @@ const login = async (req, res) => {
...
+      .status(StatusCodes.OK)
+      .json({ user: { userEmail, firstName, lastName }, token });
   } catch (error) {
     console.error('Error logging in:', error);
     const statusCode = error.statusCode || StatusCodes.INTERNAL_SERVER_ERROR;
     const errorMessage = error.message || 'Error logging in';
     res.status(statusCode).json({ message: errorMessage });
   }
- };
+};
```
Comment (benochi) [inline src/controllers/userController.js]: I know this is unrealted to this pr, but lets make sure we invalidate the session cookie: 
const logout = (req, res) => {
  try {
    res
      .clearCookie('sessionToken', { httpOnly: true, secure: true, sameSite: 'strict' }) // Adjust options as needed
      .status(StatusCodes.OK)
      .json({ message: 'Logout successful' });
  } catch (error) {
    console.error('Error during logout:', error);
    res
      .status(StatusCodes.INTERNAL_SERVER_ERROR)
      .json({ message: 'Logout failed' });
  }
};

**Review by benochi — APPROVED** [review]
approved, but see notes please.

File: src/controllers/userController.js (hunk 3 above)
Comment (Tanyaa-a) [reply src/controllers/userController.js]: Thank you Dan. We'll add it

File: src/controllers/userController.js [hunk 4]
```diff
@@ -32,35 +40,82 @@ const login = async (req, res) => {
...
+      userId,
+      { $set: updates },
+      { new: true, runValidators: true }
+    );
+
+    if (!updatedUser) {
+      throw new BadRequestError('User not found');
+    }
+
+    res.status(StatusCodes.OK).json({ user: updatedUser });
```
Comment (vrudnitskaya) [inline src/controllers/userController.js]: Tanya, great work! Please delete the password and __v from the response

**Review by vrudnitskaya — APPROVED** [review]

**Pull request merged**